File: src/systems/cardio.cpp
#include "cardio.h"

#include <optional>
#include "mkb/mkb.h"
#include "systems/heap.h"
#include "systems/log.h"
#include "systems/modlink.h"
#include "utils/draw.h"

namespace cardio {

// Corresponds to CARD call we're waiting for
enum class WriteState {
    Idle,
    Probe,
    Mount,
    Create,  // If memcard file doesn't exist
    Delete,  // Else, it exists but it's too small, delete and create from scratch
    Write,
};

struct WriteParams {
    const char* file_name;
    const void* buf;
    u32 buf_size;
    void (*callback)(mkb::CARDResult);
};

// We need a 40KB(!) buffer just for the privilege of accessing memory cards, this sucks!
// Reminder we only have ~550KB to work with for the entire mod, including savestates
static void* s_card_work_area;
static mkb::CARDFileInfo s_card_file_info;

static WriteState s_state = WriteState::Idle;
static WriteParams s_write_params;                  // Current params
static std::optional<WriteParams> s_write_request;  // Params for use for next write
static u32 s_write_size;  // Sector size of memory card A which we read when probing it

static char s_orig_gamecode[6];

/*
 * Probably not Nintendo-approved hack for letting us read/write to the same savefile even if the
 * gamecode varies. Just modify the gamecode (stored at 0x80000000) to GM2E8P (vanilla SMB2) before
 * doing memcard operations!
 */

static void set_fake_gamecode() {
    mkb::memcpy(mkb::DVD_GAME_NAME, (void*)"GM2E8P", sizeof(s_orig_gamecode));
    mkb::DCFlushRange(mkb::DVD_GAME_NAME, sizeof(s_orig_gamecode));
}

static void restore_original_gamecode() {
    mkb::memcpy(mkb::DVD_GAME_NAME, s_orig_gamecode, sizeof(s_orig_gamecode));
    mkb::DCFlushRange(mkb::DVD_GAME_NAME, sizeof(s_orig_gamecode));
}

static mkb::CARDResult read_file_internal(const char* file_name, void** out_buf) {
    mkb::CARDResult res = mkb::CARD_RESULT_READY;

    // Probe card
    do {
        res = mkb::CARDProbeEx(0, nullptr, nullptr);
    } while (res == mkb::CARD_RESULT_BUSY);
    if (res != mkb::CARD_RESULT_READY) {
        return res;
    }

    // Mount card
    mkb::CARDMountAsync(0, s_card_work_area, nullptr, nullptr);
    do {
        res = mkb::CARDGetResultCode(0);
    } while (res == mkb::CARD_RESULT_BUSY);
    if (res != mkb::CARD_RESULT_READY) {
        return res;
    }

    // Open file
    res = mkb::CARDOpen(0, const_cast<char*>(file_name), &s_card_file_info);
    if (res != mkb::CARD_RESULT_READY) {
        mkb::CARDUnmount(0);
        return res;
    }

    // Get file size
    mkb::CARDStat stat;
    res = mkb::CARDGetStatus(0, s_card_file_info.fileNo, &stat);
    if (res != mkb::CARD_RESULT_READY) {
        mkb::CARDUnmount(0);
        return res;
    }

    u32 buf_size = (stat.length + mkb::CARD_READ_SIZE - 1) & ~(mkb::CARD_READ_SIZE - 1);
    void* buf = heap::alloc(buf_size);
    if (buf == nullptr) {
        // Not quite the right error (we're out of memory, not out of card space)
        mkb::CARDUnmount(0);
        return mkb::CARD_RESULT_INSSPACE;
    }

    mkb::CARDReadAsync(&s_card_file_info, buf, buf_size, 0, nullptr);
    do {
        res = mkb::CARDGetResultCode(0);
    } while (res == mkb::CARD_RESULT_BUSY);
    if (res != mkb::CARD_RESULT_READY) {
        heap::free(buf);
        mkb::CARDUnmount(0);
        return res;
    }

    *out_buf = buf;
    return mkb::CARD_RESULT_READY;
}

mkb::CARDResult read_file(const char* file_name, void** out_buf) {
    set_fake_gamecode();
    mkb::CARDResult res = read_file_internal(file_name, out_buf);
    restore_original_gamecode();
    return res;
}

void write_file(const char* file_name, const void* buf, u32 buf_size,
                void (*callback)(mkb::CARDResult)) {
    s_write_request.emplace(WriteParams{file_name, buf, buf_size, callback});
}

static bool connect_shared_work_area() {
    if (modlink::get() == nullptr) return false;
    if (modlink::get()->modlink_version.minor < 1) return false;
    if (modlink::get()->part2->card_work_area == nullptr) return false;
    s_card_work_area = modlink::get()->part2->card_work_area;
    return true;
}

void init() {
    mkb::memcpy(s_orig_gamecode, mkb::DVD_GAME_NAME, sizeof(s_orig_gamecode));
    if (!connect_shared_work_area()) {
        s_card_work_area = heap::alloc(mkb::CARD_WORKAREA_SIZE);
    }
}

static void finish_write(mkb::CARDResult res) {
    mkb::CARDUnmount(0);  // I'm assuming that trying to unmount when mounting failed is OK
    s_write_params.callback(res);
    s_state = WriteState::Idle;
    restore_original_gamecode();
}

static void tick_state_machine() {
    mkb::CARDResult res;

    switch (s_state) {
        case WriteState::Idle: {
            if (!s_write_request.has_value()) {
                break;
            }

            // Kick off write operation
            s_write_params = s_write_request.value();
            s_write_request.reset();
            set_fake_gamecode();

            // Probe and begin mounting card A
            s_state = WriteState::Probe;
            break;
        }

        case WriteState::Probe: {
            s32 sector_size;
            res = mkb::CARDProbeEx(0, nullptr, &sector_size);
            if (res == mkb::CARD_RESULT_BUSY) {
                break;
            }
            if (res != mkb::CARD_RESULT_READY) {
                finish_write(res);
                break;
            }

            s_write_size = (s_write_params.buf_size + sector_size - 1) & ~(sector_size - 1);
            mkb::CARDMountAsync(0, s_card_work_area, nullptr, nullptr);
            s_state = WriteState::Mount;
            break;
        }

        case WriteState::Mount: {
            res = mkb::CARDGetResultCode(0);
            if (res == mkb::CARD_RESULT_BUSY) {
                break;
            }
            if (res != mkb::CARD_RESULT_READY) {
                finish_write(res);
                break;
            }

            // Try to open the file
            res = mkb::CARDOpen(0, const_cast<char*>(s_write_params.file_name), &s_card_file_info);
            if (res == mkb::CARD_RESULT_READY) {
                // Check if file is too small
                mkb::CARDStat stat;
                res = mkb::CARDGetStatus(0, s_card_file_info.fileNo, &stat);
                if (res != mkb::CARD_RESULT_READY) {
                    finish_write(res);

                } else if (stat.length < s_write_size) {
                    // Recreate file
                    mkb::CARDFastDeleteAsync(0, s_card_file_info.fileNo, nullptr);
                    s_state = WriteState::Delete;

                } else {
                    // Card opened successfully, proceed directly to writing
                    mkb::CARDWriteAsync(&s_card_file_info, const_cast<void*>(s_write_params.buf),
                                        s_write_size, 0, nullptr);
                    s_state = WriteState::Write;
                }

            } else if (res == mkb::CARD_RESULT_NOFILE) {
                // Create new file
                mkb::CARDCreateAsync(0, const_cast<char*>(s_write_params.file_name), s_write_size,
                                     &s_card_file_info, nullptr);
                s_state = WriteState::Create;

            } else {
                // Some other error, fail entire write operation
                finish_write(res);
            }
            break;
        }

        case WriteState::Create: {
            res = mkb::CARDGetResultCode(0);
            if (res == mkb::CARD_RESULT_BUSY) {
                break;
            }
            if (res != mkb::CARD_RESULT_READY) {
                finish_write(res);
                break;
            }

            mkb::CARDWriteAsync(&s_card_file_info, const_cast<void*>(s_write_params.buf),
                                s_write_size, 0, nullptr);
            s_state = WriteState::Write;
            break;
        }

        case WriteState::Delete: {
            res = mkb::CARDGetResultCode(0);
            if (res == mkb::CARD_RESULT_BUSY) {
                break;
            }
            if (res != mkb::CARD_RESULT_READY) {
                finish_write(res);
                break;
            }
            mkb::CARDCreateAsync(0, const_cast<char*>(s_write_params.file_name), s_write_size,
                                 &s_card_file_info, nullptr);
            s_state = WriteState::Create;
            break;
        }

        case WriteState::Write: {
            res = mkb::CARDGetResultCode(0);
            if (res != mkb::CARD_RESULT_BUSY) {
                // Either succeeded or failed, either way we're done
                finish_write(res);
            }
            break;
        }
    }
}

void tick() {
    WriteState prev_state;
    do {
        prev_state = s_state;
        tick_state_machine();
    } while (prev_state != s_state);
}

}  // namespace cardio

File: src/systems/cardio.rs
extern crate alloc;

use core::ffi::c_char;
use core::ffi::c_long;
use core::ffi::c_void;
use core::ptr::null_mut;

use crate::mkb;
use crate::mkb_suppl::to_card_result;
use crate::mkb_suppl::CARDResult;
use crate::mkb_suppl::CARD_READ_SIZE;
use crate::mkb_suppl::CARD_WORKAREA_SIZE;
use crate::new_cstr;
use crate::utils::math;
use crate::utils::math::round_up_pow2;
use crate::utils::modlink::ModLink;
use alloc::vec;
use alloc::vec::Vec;
use arrayvec::ArrayString;

struct FakeGamecode {
    orig_game_name: [c_char; 4],
    orig_game_company: [c_char; 2],
}

impl FakeGamecode {
    fn new() -> Self {
        unsafe {
            mkb::DVD_GAME_NAME = [b'G' as i8, b'M' as i8, b'2' as i8, b'E' as i8];
            mkb::DVD_COMPANY = [b'8' as i8, b'P' as i8];

            Self {
                orig_game_name: mkb::DVD_GAME_NAME,
                orig_game_company: mkb::DVD_COMPANY,
            }
        }
    }
}

impl Drop for FakeGamecode {
    fn drop(&mut self) {
        unsafe {
            mkb::DVD_GAME_NAME = self.orig_game_name;
            mkb::DVD_COMPANY = self.orig_game_company;
        }
    }
}

#[derive(PartialEq, Eq, Clone, Copy)]
enum WriteState {
    Probe,
    Mount,
    Create,
    Delete,
    Write,
    Done,
}

struct WriteRequest {
    file_name: ArrayString<16>,
    buf: Vec<u8>,
    write_size: usize,
    state: WriteState,
    // Sets fake gamecube when request started, unsets when completed
    fake_gamecode: FakeGamecode,
    card_result: CARDResult,
}

pub struct CardIo {
    card_work_area: &'static mut [u8],
    card_file_info: mkb::CARDFileInfo,
    write_request: Option<WriteRequest>,
}

impl CardIo {
    pub fn new(modlink: &ModLink) -> Self {
        let card_work_area = unsafe {
            match modlink.card_work_area {
                Some(work_area) => {
                    core::slice::from_raw_parts_mut(work_area as *mut u8, CARD_WORKAREA_SIZE)
                }
                None => vec![0; CARD_WORKAREA_SIZE].leak(),
            }
        };
        Self {
            card_work_area,
            card_file_info: mkb::CARDFileInfo::default(),
            write_request: None,
        }
    }

    // Synchronous at the moment. Also, do not call while write_file() is running!
    pub fn read_file(&mut self, file_name: &str) -> Result<Vec<u8>, CARDResult> {
        unsafe {
            let _fake_gamecode = FakeGamecode::new();
            let mut res;

            // Probe card
            loop {
                res = to_card_result(mkb::CARDProbeEx(0, null_mut(), null_mut()));
                if res != CARDResult::Busy {
                    break;
                }
            }
            if res != CARDResult::Ready {
                return Err(res);
            }

            // Mount card
            mkb::CARDMountAsync(
                0,
                self.card_work_area.as_mut_ptr() as *mut c_void,
                null_mut(),
                null_mut(),
            );
            loop {
                res = to_card_result(mkb::CARDGetResultCode(0));
                if res != CARDResult::Busy {
                    break;
                }
            }
            if res != CARDResult::Ready {
                return Err(res);
            }
            // Open file
            res = to_card_result(mkb::CARDOpen(
                0,
                new_cstr!(file_name, 16),
                &mut self.card_file_info,
            ));
            if res != CARDResult::Ready {
                mkb::CARDUnmount(0);
                return Err(res);
            }

            // Get file size
            let mut stat = mkb::CARDStat::default();
            res = to_card_result(mkb::CARDGetStatus(0, self.card_file_info.fileNo, &mut stat));
            if res != CARDResult::Ready {
                mkb::CARDUnmount(0);
                return Err(res);
            }

            let buf_size = math::round_up_pow2(stat.length as usize, CARD_READ_SIZE as usize);
            let mut buf = vec![0u8; buf_size];

            mkb::CARDReadAsync(
                &mut self.card_file_info as *mut _,
                buf.as_mut_ptr() as *mut _,
                buf_size as c_long,
                0,
                null_mut(),
            );
            loop {
                res = to_card_result(mkb::CARDGetResultCode(0));
                if res != CARDResult::Busy {
                    break;
                }
            }
            if res != CARDResult::Ready {
                mkb::CARDUnmount(0);
                return Err(res);
            }

            mkb::CARDUnmount(0);
            Ok(buf)
        }
    }

    // Writes asynchronously.
    // Takes ownership of the write buffer and gives it back when the write is complete
    pub fn begin_write_file(&mut self, file_name: &str, buf: Vec<u8>) {
        assert!(matches!(self.write_request, None));
        self.write_request = Some(WriteRequest {
            file_name: ArrayString::from(file_name).unwrap(),
            buf,
            write_size: 0,
            state: WriteState::Probe,
            fake_gamecode: FakeGamecode::new(),
            card_result: CARDResult::Ready,
        });
    }

    // Returns the written buffer back and potential error when completed, or None if still writing
    // or not writing at all
    pub fn get_write_result(&mut self) -> Option<(Vec<u8>, CARDResult)> {
        // We could avoid matching twice by taking then matching, but that would require moving the
        // matched write request back into self.write_request if not done
        if let Some(req) = self.write_request.as_ref() {
            if req.state == WriteState::Done {
                if let Some(req) = self.write_request.take() {
                    return Some((req.buf, req.card_result));
                }
            }
        }
        None
    }

    pub fn tick(&mut self) {
        if let Some(req) = self.write_request.as_ref() {
            let mut prev_state = req.state;
            loop {
                self.tick_state_machine();
                let curr_state = self.write_request.as_ref().unwrap().state;
                if curr_state == prev_state {
                    break;
                }
                prev_state = curr_state;
            }
        }
    }

    fn finish_write(req: &mut WriteRequest, res: CARDResult) {
        unsafe {
            mkb::CARDUnmount(0); // I'm assuming that trying to unmount when mounting failed is OK
            req.card_result = res;
            req.state = WriteState::Done;
        }
    }

    fn tick_state_machine(&mut self) {
        let req = self.write_request.as_mut().unwrap();

        unsafe {
            match req.state {
                WriteState::Probe => {
                    let mut sector_size: c_long = 0;
                    let res = to_card_result(mkb::CARDProbeEx(
                        0,
                        core::ptr::null_mut(),
                        &mut sector_size,
                    ));
                    if res == CARDResult::Busy {
                        return;
                    }
                    if res != CARDResult::Ready {
                        Self::finish_write(req, res);
                        return;
                    }

                    req.write_size =
                        (req.buf.len() + sector_size as usize - 1) & !(sector_size as usize - 1);
                    mkb::CARDMountAsync(
                        0,
                        self.card_work_area.as_mut_ptr() as *mut c_void,
                        core::ptr::null_mut(),
                        core::ptr::null_mut(),
                    );
                    req.state = WriteState::Mount;
                }

                WriteState::Mount => {
                    let res = to_card_result(mkb::CARDGetResultCode(0));
                    if res == CARDResult::Busy {
                        return;
                    }
                    if res != CARDResult::Ready {
                        Self::finish_write(req, res);
                        return;
                    }

                    // Try to open the file
                    let res = to_card_result(mkb::CARDOpen(
                        0,
                        new_cstr!(req.file_name.as_str(), 16),
                        &mut self.card_file_info,
                    ));
                    if res == CARDResult::Ready {
                        // Check if file is too small
                        let mut stat = mkb::CARDStat::default();
                        let res = to_card_result(mkb::CARDGetStatus(
                            0,
                            self.card_file_info.fileNo,
                            &mut stat,
                        ));
                        if res != CARDResult::Ready {
                            Self::finish_write(req, res);
                        } else if (stat.length as usize) < req.write_size {
                            // Recreate file
                            mkb::CARDFastDeleteAsync(
                                0,
                                self.card_file_info.fileNo,
                                core::ptr::null_mut(),
                            );
                            req.state = WriteState::Delete;
                        } else {
                            // Card opened successfully, proceed directly to writing
                            mkb::CARDWriteAsync(
                                &mut self.card_file_info,
                                req.buf.as_ptr() as *mut c_void,
                                req.write_size as c_long,
                                0,
                                core::ptr::null_mut(),
                            );
                            req.state = WriteState::Write;
                        }
                    } else if res == CARDResult::NoFile {
                        // Create new file
                        mkb::CARDCreateAsync(
                            0,
                            new_cstr!(req.file_name.as_str(), 16),
                            req.write_size as u32,
                            &mut self.card_file_info,
                            core::ptr::null_mut(),
                        );
                        req.state = WriteState::Create;
                    } else {
                        // Some other error, fail entire write operation
                        Self::finish_write(req, res);
                    }
                }

                WriteState::Create => {
                    let res = to_card_result(mkb::CARDGetResultCode(0));
                    if res == CARDResult::Busy {
                        return;
                    }
                    if res != CARDResult::Ready {
                        Self::finish_write(req, res);
                        return;
                    }

                    mkb::CARDWriteAsync(
                        &mut self.card_file_info,
                        req.buf.as_ptr() as *mut c_void,
                        req.write_size as c_long,
                        0,
                        core::ptr::null_mut(),
                    );
                    req.state = WriteState::Write;
                }

                WriteState::Delete => {
                    let res = to_card_result(mkb::CARDGetResultCode(0));
                    if res == CARDResult::Busy {
                        return;
                    }
                    if res != CARDResult::Ready {
                        Self::finish_write(req, res);
                        return;
                    }
                    mkb::CARDCreateAsync(
                        0,
                        new_cstr!(req.file_name.as_str(), 16),
                        req.write_size as u32,
                        &mut self.card_file_info,
                        core::ptr::null_mut(),
                    );
                    req.state = WriteState::Create;
                }

                WriteState::Write => {
                    let res = to_card_result(mkb::CARDGetResultCode(0));
                    if res != CARDResult::Busy {
                        // Either succeeded or failed, either way we're done
                        Self::finish_write(req, res);
                    }
                }

                _ => {}
            }
        }
    }
}
